src/MPDeviceBleImpl.cpp

```diff
@@ -0,0 +1,216 @@
+#include "MPDeviceBleImpl.h"
+#include "AsyncJobs.h"
+#include "MessageProtocol/MessageProtocolBLE.h"
+
+MPDeviceBleImpl::MPDeviceBleImpl(MessageProtocolBLE* mesProt, MPDevice *dev):
+    bleProt(mesProt),
+    mpDev(dev)
+{
+
+}
+
+bool MPDeviceBleImpl::isFirstPacket(const QByteArray &data)
+{
+    quint8 actPacket = (data[1] & 0xF0) >> 4;
+    return actPacket == 0;
+}
+
+bool MPDeviceBleImpl::isLastPacket(const QByteArray &data)
+{
+    quint8 totalPacketNum = data[1] & 0x0F;
+    quint8 actPacket = (data[1] & 0xF0) >> 4;
+    return actPacket == totalPacketNum;
+}
+
+void MPDeviceBleImpl::getPlatInfo(const MessageHandlerCbData &cb)
+{
+    auto *jobs = new AsyncJobs("Get PlatInfo", mpDev);
+
+    jobs->append(new MPCommandJob(mpDev, MPCmd::CMD_DBG_GET_PLAT_INFO, bleProt->getDefaultFuncDone()));
+
+    connect(jobs, &AsyncJobs::finished, [this, cb](const QByteArray &data)
+    {
+        Q_UNUSED(data);
+        /* Callback */
+        cb(true, "", bleProt->getFullPayload(data));
+    });
+
+    dequeueAndRun(jobs);
+}
+
+QVector<int> MPDeviceBleImpl::calcPlatInfo(const QByteArray &platInfo)
+{
+    QVector<int> platInfos;
+    platInfos.append(bleProt->toIntFromBigEndian(static_cast<quint8>(platInfo[0]), static_cast<quint8>(platInfo[1])));
+    platInfos.append(bleProt->toIntFromBigEndian(static_cast<quint8>(platInfo[2]), static_cast<quint8>(platInfo[3])));
+    platInfos.append(bleProt->toIntFromBigEndian(static_cast<quint8>(platInfo[64]), static_cast<quint8>(platInfo[65])));
+    platInfos.append(bleProt->toIntFromBigEndian(static_cast<quint8>(platInfo[66]), static_cast<quint8>(platInfo[67])));
+    return platInfos;
+}
+
+void MPDeviceBleImpl::flashMCU(QString type, const MessageHandlerCb &cb)
+{
+    if (type != "aux" && type != "main")
+    {
+        qCritical() << "Flashing called for an invalid type of " << type;
+        cb(false, "Failed flash.");
+        return;
+    }
+    auto *jobs = new AsyncJobs(QString("Flashing %1 MCU").arg(type), this);
+    const bool isAuxFlash = type == "aux";
+    const quint8 cmd = isAuxFlash ? MPCmd::CMD_DBG_FLASH_AUX_MCU : MPCmd::CMD_DBG_REBOOT_TO_BOOTLOADER;
+    jobs->append(new MPCommandJob(mpDev, cmd, bleProt->getDefaultFuncDone()));
+
+    connect(jobs, &AsyncJobs::failed, [cb, isAuxFlash](AsyncJob *failedJob)
+    {
+        Q_UNUSED(failedJob);
+        if (isAuxFlash)
+        {
+            cb(false, "Failed to flash Aux MCU!");
+        }
+        else
+        {
+            /**
+              * Main MCU flash does not cause a reconnect of the device,
+              * so daemon keeps sending messages to the device, which
+              * is causing failure in the communication, this is why
+              * this sleep is added.
+              */
+            qDebug() << "Waiting for device to start after Main MCU flash";
+            QThread::sleep(5);
```

Comment:
This blocks the UI....

You should append a new TimerJob, which will delay the job queue

Reply:
I tested this a little bit, and because it is on daemon's thread the UI was not blocked.
But I wasn't aware of TimerJob, it is a much clearer solution. And solves one more problem, because I can send it right after the flash command. (Do not have to wait until the flash command fails)